File: tests/locale_stub.mjs
// The locale stub's own gate: `node tests/locale_stub.mjs <page> <table>`,
// which `zig build test` runs (build.zig's addLocaleStubCheck).
//
// The stub is the one page nokre writes where a decision the library
// owns is transcribed into JavaScript: `Bundle.resolve` cannot be called
// from a page that loads no wasm, so locale_stub.js repeats it. This
// file is what keeps the repeat honest — for every device tag in the
// table, the *page's own* script is executed against a browser that
// reports that tag, and where it navigates is compared with the answer
// `L.resolve` gave in Zig (tests/locale_stub.zig wrote both files).
//
// So nothing here states what the right locale is. The Zig side states
// it, the page decides it, and this compares them.

import assert from "node:assert/strict";
import fs from "node:fs/promises";
import vm from "node:vm";

const [pagePath, tablePath] = process.argv.slice(2);
if (!pagePath || !tablePath) {
  console.error("usage: node tests/locale_stub.mjs <page.html> <table.json>");
  process.exit(2);
}

const page = await fs.readFile(pagePath, "utf8");
const table = JSON.parse(await fs.readFile(tablePath, "utf8"));

// The script as the browser gets it: the bytes between the tags, not a
// re-import of the source file. A stub whose emitter mangled the data,
// or wrote the call wrong, fails here.
const opened = page.indexOf("<script>");
const closed = page.indexOf("</script>", opened);
assert.ok(opened !== -1 && closed !== -1, "the stub carries no script");
const script = page.slice(opened + "<script>".length, closed);

const origin = "https://nokre.test";

/// One load of the stub: a browser reporting `language`, standing at
/// `at`. Returns where the page sent it, or null if it stayed.
function load(language, at) {
  const url = new URL(at);
  let went = null;
  const location = {
    href: url.href,
    search: url.search,
    hash: url.hash,
    replace(to) {
      went = to;
    },
  };
  vm.runInNewContext(script, { navigator: { language }, location, URL });
  return went;
}

// ---- the resolution, against the bundle's own answers ----------------

for (const { language, href } of table.cases) {
  const went = load(language, `${origin}/docs/`);
  assert.equal(
    went,
    origin + href,
    `a browser reporting ${JSON.stringify(language)} was sent to ${went}, ` +
      `where L.resolve says ${origin + href}`,
  );
}

// ---- and the pass a small bundle cannot reach -----------------------
//
// `resolve`'s middle pass is bare language **in bundle order**, and with
// no two locales sharing a language it has one candidate — so a
// transcription that scanned backwards, or that preferred the bare tag,
// would pass every assertion above. The branch is only tested if the
// bundle can pose the question, and that is asserted here rather than
// assumed: a bundle that stopped posing it fails loudly instead of
// leaving the one line most likely to drift uncovered.

const norm = (t) => t.replace(/[A-Z]/g, (u) => u.toLowerCase()).replace(/-/g, "_");
const language = (t) => norm(t).split("_")[0];
const exact = (t) => table.tags.some((u) => norm(u) === norm(t));

const contested = table.cases.find(
  (c) =>
    !exact(c.language) &&
    table.tags.filter((t) => language(t) === language(c.language)).length > 1,
);
assert.ok(
  contested,
  "no device tag in the table reaches the bundle-order pass with more than " +
    "one candidate — the branch is untested, whatever the assertions above say",
);

// What Zig answered for it *is* the earliest candidate: the rule stated,
// not just the answer copied. A reordered ARB list that quietly made the
// two candidates equivalent fails here.
const earliest =
  table.hrefs[table.tags.findIndex((t) => language(t) === language(contested.language))];
assert.equal(
  contested.href,
  earliest,
  `L.resolve sent ${JSON.stringify(contested.language)} to ${contested.href}, ` +
    `which is not the earliest bundled locale of that language (${earliest})`,
);

// ---- what the page owes the reader ----------------------------------

// The no-JavaScript half is the document itself: every locale's copy is
// a plain link, so a reader whose script never ran is offered a choice
// rather than left on a blank page.
for (const href of table.hrefs) {
  assert.ok(
    page.includes(`href="${href}"`),
    `no plain link to ${href} behind the script`,
  );
}
for (const tag of table.tags) {
  assert.ok(page.includes(`hreflang="${tag}"`), `no link marked ${tag}`);
}

// The query and the fragment are the reader's. A shared `/docs/#seams`
// has to arrive at `/{locale}/docs/#seams`, or the link lost what it
// named. The destination is the table's, so this cannot drift from the
// bundle either.
assert.equal(
  load(contested.language, `${origin}/docs/?q=1#seams`),
  `${origin}${contested.href}?q=1#seams`,
);

// A stub standing at one of its own choices is a misconfiguration, and
// the answer to it is a page that stands still rather than a browser
// that spins. Once per locale, each at its own address, with a tag that
// resolves there.
for (const href of table.hrefs) {
  const arrives = table.cases.find((c) => c.href === href);
  assert.ok(arrives, `no device tag in the table resolves to ${href}`);
  assert.equal(load(arrives.language, origin + href), null);
}

console.error(
  `locale stub: ${table.cases.length} device tags, ${table.tags.length} locales — all ok`,
);

File: tests/locale_stub.zig
//! The Zig half of the locale stub's gate: writes one real stub page
//! and, beside it, the answer `L.resolve` gives for every device tag the
//! JavaScript in that page will be asked about.
//!
//! It exists because the stub is the one place in this library where a
//! decision nokre owns has to be *transcribed* into another language.
//! `Bundle.resolve` is comptime Zig; the page that has to make the same
//! decision loads no wasm, on purpose — a redirect that first fetches an
//! app is a redirect nobody waits for. So locale_stub.js repeats the
//! algorithm, and a repeat with no gate under it is two policies waiting
//! to disagree.
//!
//! The shape is `tests/web_services.zig`'s: nothing here asserts
//! anything. It produces the two files, and tests/locale_stub.mjs runs
//! the page's own script against the table — so what is compared is the
//! bytes that ship, not a restatement of them.
//!
//! `zig build test` runs both (build.zig's `addLocaleStubCheck`).

const std = @import("std");
const nok = @import("nokre");

const dom = nok.render.dom;

comptime {
    // A program that links the library owes the hooks a shell owes
    // (docs/testing.md, "Where the harness stops").
    _ = nok.testing.shell;
}

/// **Three locales, and the order of the last two is the point.**
///
/// `resolve`'s three passes are exact tag, then **bare language in
/// bundle order**, then the template. A two-locale `en`/`fa` bundle
/// cannot reach the middle one in any interesting way: no two locales
/// share a language, so its loop has one candidate and `findIndex`,
/// `findLast` and an object's own key order all answer alike. The one
/// branch where the transcription could silently disagree would then be
/// the one branch nothing executes.
///
/// So `fa-AF` is listed **before** `fa`, and a reader asking for
/// `fa-IR` has two candidates and must land on the *earlier* — which is
/// deliberately the answer a JavaScript that scanned backwards, or that
/// preferred the bare tag, would get wrong. Order decides, and nothing
/// else does.
///
/// It is a bundle of its own rather than the web harness's pair: that
/// one is the locale service's gate and is about something else, and
/// giving it a third locale would change a test to serve this one.
/// Here the catalog carries exactly one key, the name each language
/// calls itself — which is what the stub's links say, so the labels
/// below come out of the catalog instead of being typed twice.
const L = nok.l10n.Bundle(&.{
    @embedFile("l10n/stub_en.arb"),
    @embedFile("l10n/stub_fa_AF.arb"),
    @embedFile("l10n/stub_fa.arb"),
});

/// Where the stub's choices point: paths that differ only in their
/// locale segment, which is what the prefix-all scheme produces — and
/// which nokre had no hand in, since the path scheme is the driver's
/// whole.
fn href(loc: L.Locale) []const u8 {
    return switch (loc) {
        .en => "/en/docs/",
        .fa_AF => "/fa-AF/docs/",
        .fa => "/fa/docs/",
    };
}

/// Device tags a browser can report, chosen for every answer `resolve`
/// has and for the seam between them.
///
/// The *expected* answer is deliberately not written here. Whatever
/// `L.resolve` says is the truth this gate holds the JavaScript to, so a
/// change to the resolution rule moves both sides at once and a change
/// to only one of them fails.
const probes = [_][]const u8{
    // Exact, in both cases and both separators.
    "en",
    "fa",
    "fa-AF",
    "FA",
    "EN",
    "fa_AF",
    "FA_af",
    "Fa-Af",
    // Bare-language matches with **two** candidates, where only bundle
    // order can decide and the earlier — `fa-AF` — has to win.
    "fa-IR",
    "fa_IR",
    "fa-Arab",
    "fa-Arab-IR",
    "FA-ir",
    "fa-",
    // Bare-language matches with one candidate, and region and script
    // subtags under the template.
    "en-US",
    "en_GB",
    "en-Latn-US",
    // Languages the bundle does not carry, which are the template.
    "de",
    "de-DE",
    "az-Arab",
    "zh-Hant",
    "pt-BR",
    "eng",
    "f",
    "x",
    // What a browser with nothing to say produces.
    "",
    "-fa",
};

pub fn main(init: std.process.Init) !void {
    const gpa = init.arena.allocator();
    const io = init.io;

    var args = try std.process.Args.Iterator.initAllocator(init.minimal.args, gpa);
    defer args.deinit();
    _ = args.skip();
    const page_path = args.next() orelse return error.MissingPageArgument;
    const table_path = args.next() orelse return error.MissingTableArgument;

    var app = try nok.App.init(gpa, .{ .viewport = .{ .w = 1280, .h = 1024 } });
    defer app.deinit();

    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(gpa);
    var em: dom.Emitter = .{ .gpa = gpa, .app = &app, .out = &out };
    defer em.deinit();

    try dom.localeStub(&em, L, .{
        .title = "nokre",
        .stylesheet = "/style.css",
        .heading = "Choose a language",
        .choices = .{
            .en = .{ .href = href(.en), .label = L.tr(.en, .language) },
            .fa_AF = .{ .href = href(.fa_AF), .label = L.tr(.fa_AF, .language) },
            .fa = .{ .href = href(.fa), .label = L.tr(.fa, .language) },
        },
    });

    // The table: one row per probe, the href *the bundle* sends that
    // reader to. Both sides read the destination out of `href` above, so
    // nothing here can agree with the page by accident.
    var rows: [probes.len]Row = undefined;
    for (probes, 0..) |tag, i| {
        rows[i] = .{ .language = tag, .href = href(L.resolve(tag)) };
    }

    const locales = comptime std.enums.values(L.Locale);
    var tags: [locales.len][]const u8 = undefined;
    var hrefs: [locales.len][]const u8 = undefined;
    inline for (locales, 0..) |loc, i| {
        tags[i] = comptime L.tag(loc);
        hrefs[i] = href(loc);
    }

    const table = try std.json.Stringify.valueAlloc(gpa, .{
        .tags = tags[0..],
        .hrefs = hrefs[0..],
        .cases = rows[0..],
    }, .{});

    const cwd: std.Io.Dir = .cwd();
    try cwd.writeFile(io, .{ .sub_path = page_path, .data = out.items });
    try cwd.writeFile(io, .{ .sub_path = table_path, .data = table });
}

const Row = struct { language: []const u8, href: []const u8 };
